Yank/mpi.py

```diff
@@ -0,0 +1,421 @@
+    7
+
+    """
+    broadcast_result = kwargs.pop('broadcast_result', False)
+    sync_nodes = kwargs.pop('sync_nodes', False)
+    result = None
+    mpicomm = get_mpicomm()
+
+    if mpicomm is not None:
+        node_name = 'Node {}/{}'.format(mpicomm.rank+1, mpicomm.size)
+    else:
+        node_name = 'Single node'
+
+    # Execute the task only on the specified node.
+    if mpicomm is None or mpicomm.rank == rank:
+        logger.debug('{}: executing {}'.format(node_name, task))
+        result = task(*args, **kwargs)
+
+    # Broadcast the result if required.
+    if mpicomm is not None:
+        if broadcast_result is True:
+            logger.debug('{}: waiting for broadcast of {}'.format(node_name, task))
+            result = mpicomm.bcast(result, root=rank)
+        elif sync_nodes is True:
+            logger.debug('{}: waiting for barrier after {}'.format(node_name, task))
+            mpicomm.barrier()
+
+    # Return result.
+    return result
+
+
+def on_single_node(rank, broadcast_result=False, sync_nodes=False):
+    """A decorator version of run_single_node.
+
+    Decorates a function to be always executed with run_single_node.
+
+    Parameters
+    ----------
+    rank : int
+        The rank of the MPI communicator that must execute the task.
+    broadcast_result : bool, optional
+        If True the result is broadcasted to all nodes. If False,
+        only the node executing the function will receive its return
+        value, and all other nodes will receive None (default is False).
+    sync_nodes : bool, optional
+        If True, the nodes will be synchronized at the end of the
+        execution (i.e. the task will be blocking) even if the
+        result is not broadcasted (default is False).
+
+    See Also
+    --------
+    run_single_node
+
+    Examples
+    --------
+    >>> @on_single_node(rank=0, broadcast_result=True)
+    ... def add(a, b):
+    ...     return a + b
+    >>> add(3, 4)
+    7
+
+    """
+    def _on_single_node(task):
+        @wraps_py2(task)
+        def _wrapper(*args, **kwargs):
+            kwargs['broadcast_result'] = broadcast_result
+            kwargs['sync_nodes'] = sync_nodes
+            return run_single_node(rank, task, *args, **kwargs)
+        return _wrapper
+    return _on_single_node
+
+
+def distribute(task, distributed_args, *other_args, **kwargs):
+    """Map the task on a sequence of arguments to be executed on different nodes.
+
+    If MPI is not activated, this simply runs serially on this node. The
+    algorithm guarantees that each node will be assigned to the same job_id
+    (i.e. the index of the argument in distributed_args) every time.
+
+    Parameters
+    ----------
+    task : callable
+        The task to be distributed among nodes. The task will be called as
+        task(distributed_args[job_id], *other_args, **kwargs), so the parameter
+        to be distributed must the the first one.
+    distributed_args : iterable
```

Comment:
It works for single arg items, but the "task" function can take as first argument a tuple/list, so you can still use this to distribute sets of arguments.